solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusExporter.java

```diff
@@ -0,0 +1,203 @@
+      GaugeSnapshot.GaugeDataPointSnapshot dataPoint =
+          createGaugeDatapoint(((Number) dropwizardMetric).doubleValue(), labels);
+      collectGaugeDatapoint(metricName, dataPoint);
+    } else if (dropwizardMetric instanceof HashMap) {
+      HashMap<?, ?> itemsMap = (HashMap<?, ?>) dropwizardMetric;
+      for (Object item : itemsMap.keySet()) {
+        if (itemsMap.get(item) instanceof Number) {
+          GaugeSnapshot.GaugeDataPointSnapshot dataPoint =
+              createGaugeDatapoint(
+                  ((Number) itemsMap.get(item)).doubleValue(),
+                  labels.merge(Labels.of("item", (String) item)));
+          collectGaugeDatapoint(metricName, dataPoint);
+        }
+      }
+    }
+  }
+
+  /**
+   * Create a {@link io.prometheus.metrics.model.snapshots.CounterSnapshot.CounterDataPointSnapshot}
+   * with labels
+   *
+   * @param value metric value
+   * @param labels set of name/values labels
+   */
+  public CounterSnapshot.CounterDataPointSnapshot createCounterDatapoint(
+      double value, Labels labels) {
+    return CounterSnapshot.CounterDataPointSnapshot.builder().value(value).labels(labels).build();
+  }
+
+  /**
+   * Create a {@link io.prometheus.metrics.model.snapshots.GaugeSnapshot.GaugeDataPointSnapshot}
+   * with labels
+   *
+   * @param value metric value
+   * @param labels set of name/values labels
+   */
+  public GaugeSnapshot.GaugeDataPointSnapshot createGaugeDatapoint(double value, Labels labels) {
+    return GaugeSnapshot.GaugeDataPointSnapshot.builder().value(value).labels(labels).build();
+  }
+
+  /**
+   * Collects {@link io.prometheus.metrics.model.snapshots.CounterSnapshot.CounterDataPointSnapshot}
+   * and appends to existing metric or create new metric if name does not exist
+   *
+   * @param metricName Name of metric
+   * @param dataPoint Counter datapoint to be collected
+   */
+  public void collectCounterDatapoint(
+      String metricName, CounterSnapshot.CounterDataPointSnapshot dataPoint) {
+    if (!metricCounters.containsKey(metricName)) {
+      metricCounters.put(metricName, new ArrayList<>());
+    }
+    metricCounters.get(metricName).add(dataPoint);
+  }
+
+  /**
+   * Collects {@link io.prometheus.metrics.model.snapshots.GaugeSnapshot.GaugeDataPointSnapshot} and
+   * appends to existing metric or create new metric if name does not exist
+   *
+   * @param metricName Name of metric
+   * @param dataPoint Gauge datapoint to be collected
+   */
+  public void collectGaugeDatapoint(
+      String metricName, GaugeSnapshot.GaugeDataPointSnapshot dataPoint) {
+    if (!metricGauges.containsKey(metricName)) {
+      metricGauges.put(metricName, new ArrayList<>());
+    }
+    metricGauges.get(metricName).add(dataPoint);
+  }
+
+  /**
+   * Returns an immutable {@link MetricSnapshots} from the {@link
+   * io.prometheus.metrics.model.snapshots.DataPointSnapshot}s collected from the registry
+   */
+  public MetricSnapshots collect() {
```

Comment:
Does the order matter, like will we return all counters then all gauges in the result?  If so, I could see that being kind of annoying but not a deal breaker of course.

Reply:
Order doesn't matter. The constructor of `MetricSnapshots` gets sorted by the Metric name. See this [constructor](https://github.com/prometheus/client_java/blob/c0a38270c021e4d63998a98a68bacf2b6f32ace5/prometheus-metrics-model/src/main/java/io/prometheus/metrics/model/snapshots/MetricSnapshots.java#L40). But with metrics, it shouldn't matter anyways although I agree, it can be annoying. All the metrics would be aggregated on a Grafana level with prometheus queries.